### PR TITLE
k8s.io/apiserver/storage/etcd: refactor etcd GetList.

### PR DESCRIPTION
**Pull request by siyuanfoundation**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
refactor etcd GetList.
to make the GetList easier to understand.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by vaibhav2107** [comment]
@siyuanfoundation In order to proceed with the PR you have to sign the CLA first, Please see [here ](https://github.com/kubernetes/community/blob/master/CLA.md)for more details.

**Comment by logicalhan** [comment]
This integration test failure looks relevant: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/121049/pull-kubernetes-integration/1710861508126183424

File: test/integration/apiserver/apiserver_test.go [hunk 1]
```diff
@@ -527,12 +527,6 @@ func testListOptionsCase(t *testing.T, rsClient appsv1.ReplicaSetInterface, watc
 		}
 		return
 	}
-	if opts.ResourceVersion == invalidResourceVersion {
```
Comment (logicalhan) [inline test/integration/apiserver/apiserver_test.go:530]: I don't understand.. why is this move necessary?

Comment (siyuanfoundation) [reply test/integration/apiserver/apiserver_test.go:530]: the old code checks validity of opts.ResourceVersion before continuation, the new code checks the validity of continuation first. The order difference would result in different err message for the test case of continuation along with an invalid opts.rv

**Comment by siyuanfoundation** [comment]
> Nice job, I have been interested in the same area #120265 #120424
> 
> My recommendation is to split this PR into smaller ones. Extract a trivial code change, so the review takes seconds and not minutes as most people that can review/approve this code are overloaded with review.
> 
> For example extraction of `getCurrentState` is a good candidate to be a separate PR.

Split the original PR to:
https://github.com/kubernetes/kubernetes/pull/121049
https://github.com/kubernetes/kubernetes/pull/121158
https://github.com/kubernetes/kubernetes/pull/121159
https://github.com/kubernetes/kubernetes/pull/121160
https://github.com/kubernetes/kubernetes/pull/121161

**Comment by wojtek-t** [comment]
/lgtm
/approve

Thanks!

**Pull request merged**